### PR TITLE
ENH: Added `notes` command for spin

### PR DESCRIPTION
**Pull request by ganesh-k13**
### Changes
- Added a notes command for spin
- Created `main` to support imports in `test_all_newsfragments_used.py`

### Testing
</details>

<details>
  <summary> Generate notes with version</summary>

```py
~/os/numpy (bld_24080_notes) » spin notes --version-override 2.0                                                                                                                     ganesh@ganesh-MS-7B86
Generating release notes for NumPy 2.0
Loading template...
Finding news fragments...
Rendering news fragments...
Writing to newsfile...
Staging newsfile...
Removing news fragments...
Removing the following files:
/home/ganesh/os/numpy/doc/release/upcoming_changes/24532.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24946.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24877.improvement.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24775.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24858.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24634.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24540.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24854.deprecation.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24680.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24770.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24555.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24445.deprecation.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24945.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24922.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24555.improvement.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24818.deprecation.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24717.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24053.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24587.python_removal.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24376.python_removal.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24887.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24931.improvement.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24477.python_removal.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24807.python_removal.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24420.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/19355.new_feature.rst
Done!

Verifying consumption of all news fragments

```
</details>

---

</details>

<details>
  <summary> Auto detect version</summary>

```py
~/os/numpy (bld_24080_notes*) » spin notes                                                                                                                                  ganesh@ganesh-MS-7B86
Using inferred version 1.26.1
Generating release notes for NumPy 1.26.1
Loading template...
Finding news fragments...
Rendering news fragments...
Writing to newsfile...
Staging newsfile...
Removing news fragments...
Removing the following files:
/home/ganesh/os/numpy/doc/release/upcoming_changes/24532.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24946.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24877.improvement.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24775.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24858.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24634.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24540.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24854.deprecation.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24680.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24770.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24555.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24445.deprecation.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24945.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24922.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24555.improvement.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24818.deprecation.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24717.change.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24053.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24587.python_removal.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24376.python_removal.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24887.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24931.improvement.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24477.python_removal.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24807.python_removal.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/24420.new_feature.rst
/home/ganesh/os/numpy/doc/release/upcoming_changes/19355.new_feature.rst
Done!

Verifying consumption of all news fragments
```
</details>

---

</details>

<details>
  <summary> `towncrier` not found </summary>

```py
~/os/numpy (bld_24080_notes*) » spin notes                                                                                                                                  ganesh@ganesh-MS-7B86
Using inferred version 1.26.1
Generating release notes for NumPy 1.26.1
Error: please install `towncrier` to use this command

```
</details>

---

</details>

<details>
  <summary> Auto version detection failed</summary>

```py
~/os/numpy (bld_24080_notes*) » spin notes                                                                                                                                  ganesh@ganesh-MS-7B86
Error: `pip show numpy` returned error: WARNING: Package(s) not found: numpy

```
</details>

### Notes
Related: #24080

### CODE REVIEW

**Comment by stefanv** [comment]
Thanks for the new command @ganesh-k13. When I run it on my repo, it picks up the installed NumPy versions, instead of the repo built version. Is this intended?

**Comment by ganesh-k13** [comment]
Ah ok, I can change that. I blindly copied from: https://github.com/numpy/numpy/blob/ae1cb2d0eb40444e812b10f5b4f16c97eee13b38/.circleci/config.yml#L67

We can use the `__version__`, which might make more sense in a desktop as compared to what we have now which is good for CI.

**Comment by stefanv** [comment]
Whatever you decide is good with me, I just wanted to point it out. Let me know when you're ready, and we'll get this in.

**Comment by ganesh-k13** [comment]
@stefanv, I have made the change to use the installed NumPy version in the latest commit.

**Review by stefanv — COMMENTED** [review]
The versioning is still a bit odd: do you *ever* want to pick up the installed NumPy to determine the version? I'd imagine the only thing that matters is whether the snippets match with the version of NumPy being released.

So, I'd have made the logic as follows:
- Find the version in `pyproject.toml` (this config is already available to you via `spin`, btw)
- You know that this version matches with the snippets about to be processed.
- Use `--version` (or perhaps `--version-override`) to specify a custom version.

This way, you will never have release notes out of sync with the version of NumPy you're releasing.

**Comment by ganesh-k13** [comment]
Thanks @stefanv, it's a lot cleaner this way.

File: .spin/cmds.py [hunk 1]
```diff
@@ -567,3 +568,67 @@ def _config_openblas(blas_variant):
...
+    \b
+    To automatically pick the version
+
+    \b
+    $ spin notes
+    """
+    if not version_override:
+        version = util.get_config()['project']['version']
+    else:
+        version = version_override
```
Comment (stefanv) [inline .spin/cmds.py:598]:
```suggestion
    version = version_override or util.get_config()['project.version']
```

File: .spin/cmds.py [hunk 2]
```diff
@@ -567,3 +568,67 @@ def _config_openblas(blas_variant):
...
+    # Check if `towncrier` is installed
+    if not shutil.which("towncrier"):
+        raise click.ClickException(
+            f"please install `towncrier` to use this command"
+        )
+
+    # towncrier build --version 2.1 --yes
+    cmd = ["towncrier", "build", "--version", version, "--yes"]
+    try:
+        p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, encoding="utf-8")
```
Comment (stefanv) [inline .spin/cmds.py:614]: Consider using `_run` as elsewhere in the file.

**Review by stefanv — COMMENTED** [review]
Some minor suggestions, but overall LGTM.

File: .spin/cmds.py (hunk 1 above)
Comment (ganesh-k13) [reply .spin/cmds.py]: Modified via [123af65](https://github.com/numpy/numpy/pull/25017/commits/123af65fb1cb6b7695b58f74adf3645cbfb42f2b)

File: .spin/cmds.py (hunk 2 above)
Comment (ganesh-k13) [reply .spin/cmds.py]: Adde via [123af65](https://github.com/numpy/numpy/pull/25017/commits/123af65fb1cb6b7695b58f74adf3645cbfb42f2b)

**Comment by stefanv** [comment]
This look great, thanks. The only other issue I saw is that towncrier doesn't print the location of the output file generated, which this command could do.

If you want to address that, we can do it in a follow-up PR.

**Pull request merged**